### DISCUSSION TITLE
Berry function return and tasmota.set_timer

### DISCUSSION THREAD

**Discussion by ouinouin** [arendst/Tasmota#16677] [Q&A] [answered]
Hi, 
stepping slowly into berry,  i have this code 
this function detects if we re on the day or night , is working and returns 1 if during the day.

```
def is_daylight()
#we search the number of minutes since midnight
var daytime_minutes = tasmota.time_dump(tasmota.rtc()['local'])['hour'] * 60 + tasmota.time_dump(tasmota.rtc()['local'])['min']
#finding sunrise in minutes
var sunrise_minutes = tasmota.strptime(tasmota.cmd('Status 7')['StatusTIM']['Sunrise'] , "%H:%M")['hour'] *60 + tasmota.strptime(tasmota.cmd('Status 7')['StatusTIM']['Sunrise'] , "%H:%M")['min']
#finding sunset in minutes
var sunset_minutes = tasmota.strptime(tasmota.cmd('Status 7')['StatusTIM']['Sunset'] , "%H:%M")['hour'] *60 + tasmota.strptime(tasmota.cmd('Status 7')['StatusTIM']['Sunset'] , "%H:%M")['min']
print(daytime_minutes)
print(sunrise_minutes)
print(sunset_minutes)
var result = 0
if daytime_minutes > sunrise_minutes && daytime_minutes < sunset_minutes
result = 1
else
result = 0
end
return result
end
```
now i d like to run this with a delayed time so from the documentation, i adapted this : 
```
def loop5000()
  tasmota.set_timer(5000, is_daylight)
end
loop5000() 
```
the issue is : i dont understand how to catch the return value of the is_daylight function , i can see the print messages on the console, but on the loop5000() function i d really like to get the result of the called function inside a variable.
i tried various things like var mytimer =  tasmota.set_timer(5000, is_daylight) in the hope to get the result in the variable , but nothing.
is this a limitation of the set_timer function or do i misunderstand how this is working ?

**Comment by sfromis** [comment, marked as answer]
The return value from the function passed to `tasmota.set_timer` is not used for anything. For that you can call a function to process the value, or assign it to a global variable defined outside of the function or class scope.

Note: Your `loop5000` function is no loop, it executes only once. For repeated executions, you might use the `set_timer_modulo` helper function available on the Berry cookbook page.

**Reply by ouinouin** [reply]
thanks for clarification yes for now i had opened the loop , i used the method in the lvgl example for scanning the button , was initially looping and looking like this

def loop5000()
  tasmota.set_timer(5000, loop5000)
end
loop5000() 

thanks also for the set timer modulo hint

**Reply by ozett** [reply]
may i jump in? the set_timer_modulo helper is way above my coding skills,
but that seems the way to have a repeating intervall call.

may i ask politely how this cook-book template would be written in real code for an intervall of 5sek.?
i cannot make up a working sketch.. a shame, sorry..

does not work... where to replace the time?
```
def set_timer_modulo(delay,f,id)
  var now=tasmota.millis()
  print (now)
  tasmota.set_timer((now+delay/4+delay)/delay*delay-now, def() set_timer_modulo(delay,f,id) f() end, id)
end

set_timer_modulo(5000)
```

**Reply by sfromis** [reply]
I very much used real code when creating that sample :smile: Not sure what is unclear?

Maybe this is what you wanted.... 
It takes same parameters as `tasmota.set_timer`, see that for explanation.
`set_timer_modulo(5000, /->print('tick tock'), 'Ticker')`

This part is just calculating delay until next time, correcting for any time drift: `(now+delay/4+delay)/delay*delay-now`

The benefit of having an anonymous function is that it works as a closure, meaning that it has access to the outer parameters delay, f  and id from the original call. `def() set_timer_modulo(delay,f,id) f() end`

Other than that, it acts the same as a regular function definition. The whole concept of set_timer (including here) is that you can pass a function to be called later, here named just `f`. It could also have been assigned to a var
`var calledfunction = def() set_timer_modulo(delay,f,id) f() end`

More generally, depending on your needs, the `cron` module in Berry was added later, and is excellent for scheduling things to be called repeatedly in the future, with much greater flexibility. Unless you want the cycle to start based on "now", instead of rounded to next interval, I'd see `cron` as the primary suggestion.

**Reply by ozett** [reply]
thanks, that helped me a lot with my big c/c++ gaps. but it's crazy fun to see what else is possible. Thank you very much. I keep experimenting. everything looks great

(seems like its possible to send a function with the call of the function 😄 
<kbd>![image](https://user-images.githubusercontent.com/5937811/216837731-75c3ed16-4cdb-47d6-a630-48c6c869dfc2.png)

edit:
> l, I'd see cron as the primary suggestion.

yes, great module. i looked into this and will also play with it. but i started my experiments with a led blinking on/off within 50ms i must stick to berry. cron looks like it has no shorter intervals than 1 sek.

**Reply by sfromis** [reply]
Well, `set_timer_modulo` is also not really suitable for less than whole-second timing, due to the calculation to avoid drift. For short intervals, you also have driver callback methods like `every_50ms`

**Reply by ozett** [reply]
> I very much used real code when creating that sample 😄 Not sure what is unclear?

I'm trying to understand how amateurish my question is for a mature programmer. 
because I can't get the function to start properly.

it seems my big misconception that i thought "scripting" was for beginners. 
but I must have gone astray, for I have read 2 times

https://tasmota.github.io/docs/Berry-Introduction/#closures
https://github.com/berry-lang/berry/wiki/Chapter-5#closure

and only read over and over "advanced" and "special case" like

``` It is sometimes seen as intimidating or complex, but it's actually very simple.```
```Advanced users: there is a compact syntax for simple callbacks: def cb(a,b) return <expr> end becomes/ a,b -> <expr>``` 
```Lambda expression is a special anonymous function```

it's not that easy to immediately land at the level of a top programmer, but i still want to believe that scripting can also be mastered by a layman. I've already started with the berry manual. the nut must be crackable (to get a repeated call of this (closure-special-case) function,) 😄

**Reply by sfromis** [reply]
Here are a few functionally equivalent ways

You already saw the short version, using abbreviated lambda syntax for the closure function to be called every 5 seconds:
```
set_timer_modulo(5000, /->print('tick tock'), 'Ticker')
```

Expanding the abbreviated syntax to have the closure as an anonymous function definition instead:
```
set_timer_modulo(5000, def () print('tick tock') end, 'Ticker')
```

You could also use a named function definition (no longer a closure) if you need no access to variables from where the function call is:
```
def ticktock()
  print('tick tock')
end
set_timer_modulo(5000, ticktock, 'Ticker')
```

An anonymous function definition can also be housed in a variable:
```
var ticktock = def () print('tick tock') end
set_timer_modulo(5000, ticktock, 'Ticker')
```

This also works with the lambda notation:
```
var ticktock = /-> print('tick tock')
set_timer_modulo(5000, ticktock, 'Ticker')
```
The lambda notation and closures is related to the concept of "functional programming", which can be a great style for making programs simpler.

Beware, this will **not** work, as the ticktock function will be called immediately, and the result passed to set_timer_modulo:
```
var ticktock = /-> print('tick tock')
set_timer_modulo(5000, ticktock(), 'Ticker')
```

**Reply by ozett** [reply]
💐 💐 💐 💐 💐 💐 💐 💐 💐 💐 💐 💐 💐 💐 
Thank you very much, Sir!  
💐 💐 💐 💐 💐 💐 💐 💐 💐 💐 💐 💐 💐 💐

**Reply by ozett** [reply]
with a little smile:
I'm a little proud of us, the led is flashing. programmed for some kind of a 50ms cycle.
```

var toggle = false

def set_timer_modulo(delay,f,id)
  tasmota.set_timer(50, def() set_timer_modulo(delay,f,id) f() end, id)
end

set_timer_modulo(50, def () tasmota.cmd('power1 ' .. (toggle)); toggle = !toggle; end, 'Ticker')

```

now i'm throwing myself into the adventure of experimenting [to frankenstein a driver](https://tasmota.github.io/docs/Berry/#creating-a-tasmota-driver) and the callback methods like `every_50ms`

**Reply by ozett** [reply]
experimenting with frankenstein variants from this skeleton. 
led looks like it seems to work with random callback-times 🤭 

```

import math
var toggle = false

def set_timer_modulo(delay,f,id)
  var interval = math.rand() % 1000 + 1000  #
  #print ('interval: ', interval)
  tasmota.set_timer(interval, def() set_timer_modulo(delay,f,id) f() end, id)
end

set_timer_modulo(10, def () tasmota.cmd('power1 ' .. (toggle)); toggle = !toggle; end, 'Ticker')

```

**Reply by sfromis** [reply]
No, that does not work, as the available timer resolution is 50 ms, numbers are basically rounded up to nearest tick, meaning max 20 Hz.

**Reply by ozett** [reply]
huh, 
so you mean there is a maximal Limit for the recurring time-value for the `interval` in the part:
``` 

tasmota.set_timer(interval, def() set_timer_modulo(delay,f,id) f() end, id)

``` 
without a warning [in the docu](https://tasmota.github.io/docs/Berry/#timers) ? thought the delay could be maxed out to hours ..

<kbd>![image](https://user-images.githubusercontent.com/5937811/217336040-bba7d20a-f1a2-4460-8106-4fff4bb10e5c.png)

or a repeated misunderstanding of concepts on my side? ... 😢

**Reply by sfromis** [reply]
No one was talking about a maximal upper limit, outside of what would overflow 31-bit numbers. I was talking about the detail that you cannot, say, have an interval of exactly 220 milliseconds, as the resolution is 50 ms. This would be rounded up to about 250 ms.

I've not tested the upper limit, but a quick calculation suggests that you should not expect a cycle time of over 24 days working.

**Reply by ozett** [reply]
in my experiments with the LED the minimum of 50ms pause seems sufficient to my eyes.

and seems better than the fixed 100ms cycle for the normal blinker-routine.  
https://github.com/arendst/Tasmota/blob/ce29bc38d753c3070479ea4897238c3a4c9f1a86/tasmota/tasmota_support/support_tasmota.ino#L1203
one could think of making also a standard-blinker with 50ms? and not only 0.1s?
https://github.com/arendst/Tasmota/blob/ce29bc38d753c3070479ea4897238c3a4c9f1a86/tasmota/tasmota_support/support_tasmota.ino#L1173

cycling with a min of 50ms and above (to 24 days) with `tasmota.set_timer()` can be done with berry and without compiling. Thats the **great news: without compiling**. But i also discovered that "berry-scripting" likes to be more a "berry-programming". but may i will frankenstein experiment this further with my limited codingskills (getting better through this). and maybe i come to try "frankenstein" a berry-driver to get down under 50ms ... tasmota is also great fun and stuff to play with. 🥇

**Reply by sfromis** [reply]
These days, "scripting" and "programming" is not a useful distinction, except that it is easier to do actual programming with a scripting-style programming environment. "Programming" does not mean compiling to machine code, and "scripting" does no longer suggest that you wrap more basic commands in a "programming light" model.

For faster than the standard 50 ms tick resolution, see https://tasmota.github.io/docs/Berry/#fast-loop

**Reply by ozett** [reply]
i am fiddling now with a repeating timer, so to say a chain of `tasmota.set_timer`.
simplest derived from from the `set_timer_modulo` example.

Learned a lot from https://github.com/berry-lang/berry/wiki/Chapter-2#222-scope-and-life-cycle
but i find myself permantenly in the stoneage of programming in ' pascal' -language.

i guess a professional polished skeleton for chained-times would be nice to have in the cookbook, too.

```
# Chained Counter
var counter = 0
var toggle = false
def set_timer_modulo(delay,f,id,c)
    print ('start function with c: ', c , ' with size:', size(c))
    if counter <  size(c)
        print (' c[',counter,']:', c[counter])

        # Switch LED on/off
        tasmota.cmd('power1 ' .. (toggle))
        toggle = !toggle

        # set next wakeup time for next timer
        delay =  c[counter]
        counter = counter + 1
    else 
        # start again 
        print ('counter reset, counter=0')
        counter = 0
        delay = 0
    end 
    tasmota.set_timer(delay, def() set_timer_modulo(delay,f,id,c) end, id)
end

# first start of the timer-chain
set_timer_modulo(5000,  def() end, 'TickerID', [4081,7101,5082,7102])
```

**Reply by ozett** [reply]
stumbled about this: the naming for [calls in this lib](https://github.com/davetcc/TaskManagerIO/) could be a naming-shematic 
for `chained` and non-blocking timer skeletons (one-shot & repeated) in berry.

```
tasmota.set_timer.scheduleOnce()
tasmota.set_timer.scheduleFixedRate()
```
or like this
```
tasmota.set_timer.oneshot()
tasmota.set_timer.repeated()
```
